### PR TITLE
planner: fix wrong PointGet / TableDual plan reused in plan cache

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23187
close https://github.com/pingcap/tidb/issues/23144
close https://github.com/pingcap/tidb/issues/23304
close https://github.com/pingcap/tidb/issues/23290

Problem Summary:

Planner cannot find a plan for a query while it indeed can.

### What is changed and how it works?

What's Changed:
- if we prune paths in `DeriveStats` and only keep the TableDual / PointGet path, we mark this plan as not cacheable;
- in skylinePruning we can enable the early return now if we find a path with empty range;

How it Works:

These plans have strong dependency on the param values, i.e, if the values change, those plans may lead to wrong query results.

### Related changes

N/A

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - This may prevent some plans which have TableDual / PointGet from being cached, while those TableDual / PointGet may have no relationship with param values at all. For example, plan for query `prepare stmt from 'select * from t1, t2 where t1.a > 9 and t1.a < 0 and t2.a > ? and t1.b = t2.b'` cannot be cached then, but it is safe actually.

### Release note 

- Fix wrong PointGet / TableDual plan reused in plan cache

### CODE REVIEW

File: executor/explainfor_test.go [hunk 1]
```diff
@@ -479,7 +479,4 @@ func (s *testPrepareSerialSuite) TestPointGetUserVarPlanCache(c *C) {
 	tk.MustQuery("execute stmt using @a").Check(testkit.Rows(
 		"2 4 2 2",
 	))
-	tk.MustQuery("select @@last_plan_from_cache").Check(testkit.Rows(
-		"1",
-	))
```
Comment (eurekaka) [inline executor/explainfor_test.go:484]: Sadly, this is the side effect of this PR, we may have false negative decision about "cache it or not", unless we track the whole range calculation procedure to record if any user variable is related to the final ranges. That would be trivial and easy to be broken by new code.

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -528,7 +528,7 @@ func (ds *DataSource) skylinePruning(prop *property.PhysicalProperty) []*candida
 			continue
 		}
 		// if we already know the range of the scan is empty, just return a TableDual
-		if len(path.Ranges) == 0 && !ds.ctx.GetSessionVars().StmtCtx.UseCache {
```
Comment (time-and-fate) [inline planner/core/find_best_task.go:531]: There's the same logic at https://github.com/pingcap/tidb/blob/5c4e8dd4e46bd7c68bc97f917472159225dc9d61/planner/core/find_best_task.go#L688
I think we should do the same change there.

**Review by time-and-fate — COMMENTED** [review]
In `(ds *DataSource).accessPathsForConds`, which is for building partial paths for index merge, there're also logics like "if we have point get or empty range, remove other paths". Seems we should also set `OptimDependOnMutableConst` to `true` there?

File: planner/core/find_best_task.go (hunk 2 above)
Comment (eurekaka) [reply planner/core/find_best_task.go:531]: We can use `OptimDependOnMutableConst` to replace `UseCache` check in other places as well, e.g, `refineArgs()`, but these changes are more like a refactor, not a bug fix, so I prefer not to do it in this PR.

File: planner/core/stats.go [hunk 3]
```diff
@@ -506,10 +508,8 @@ func (ds *DataSource) accessPathsForConds(conditions []expression.Expression, us
 				logutil.BgLogger().Debug("can not derive statistics of a path", zap.Error(err))
 				continue
 			}
-			// If `AccessConds` is empty, we ignore the access path.
-			// If the path contains a full range, ignore it also. This can happen when `AccessConds` is constant true, and
-			// it comes from the result of a subquery, so it is not folded.
-			if len(path.AccessConds) == 0 || ranger.HasFullRange(path.Ranges) {
+			// If the path contains a full range, ignore it.
+			if ranger.HasFullRange(path.Ranges) {
 				continue
```
Comment (time-and-fate) [inline planner/core/stats.go:513]: Why is this changed too?

Comment (eurekaka) [reply planner/core/stats.go:513]: It is a bug, if the filter is like `(a > 1 and a < 0) or b > 1`, for index a, both `path.AccessConds` and `path.Ranges` are empty, we should generate an `IndexMerge` plan on `TableDual`, not a `TableFullScan` plan.

File: planner/core/find_best_task.go (hunk 2 above)
Comment (time-and-fate) [reply planner/core/find_best_task.go:531]: OK. Got it.
Sometimes really hard to understand the entire logic correctly.

File: planner/core/stats.go (hunk 3 above)
Comment (time-and-fate) [reply planner/core/stats.go:513]: OK.

File: executor/explainfor_test.go (hunk 1 above)
Comment (time-and-fate) [reply executor/explainfor_test.go:484]: Can we change the test results instead of deleting these tests?
Because now we explicitly don't want to cache these plans to assure correctness.

**Review by qw4990 — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]
/lgtm

**Comment by blacktear23** [comment]
@eurekaka do we need cherry-pick to release branches?
